### PR TITLE
Initial code commit

### PR DESCRIPTION
**Pull request by phutchins**

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -0,0 +1,139 @@
+PATH
```
Comment (suyograo) [inline Gemfile.lock]: Can you remove this file? We don't checkin the lock file, just the gemfile.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+## 0.1.0
+  - Plugin created with the logstash plugin generator
```
Comment (suyograo) [inline CHANGELOG.md]: Better changelog message?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+  # The source field name which contains full path to kubelet log file.
+  config :source, :validate => :string, :default => "path"
+
+  # The target field name to write event kubernetes metadata.
+  config :target, :validate => :string, :default => "kubernetes"
+
+  # Auth for hitting the Kubernetes API. This can be either basic auth or
+  # Bearer auth. It should be formated like the following
+  # auth => {
```
Comment (suyograo) [inline lib/logstash/filters/kubernetes_metadata.rb]: This should be formatted in asciidoc. Typically, like:

```
[source, ruby]
---------------------------------
auth => {
}
---------------------------------
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  #   basic => {
+  #     user => "admin"
+  #     pass => "mysuperawesomepassword"
+  #   }
+  # }
+  #
+  # or
+  #
+  # auth => {
```
Comment (suyograo) [inline lib/logstash/filters/kubernetes_metadata.rb]: asciidoc formatting as stated above

File: logstash-filter-kubernetes_metadata.gemspec [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  # Files
+  s.files = Dir['lib/**/*','spec/**/*','vendor/**/*','*.gemspec','*.md','CONTRIBUTORS','Gemfile','LICENSE','NOTICE.TXT']
+   # Tests
+  s.test_files = s.files.grep(%r{^(test|spec|features)/})
+
+  # Special flag to let us know this is actually a logstash plugin
+  s.metadata = { "logstash_plugin" => "true", "logstash_group" => "filter" }
+
+  # Gem dependencies
+  s.add_runtime_dependency 'logstash-core-plugin-api', '~> 2.0'
```
Comment (suyograo) [inline logstash-filter-kubernetes_metadata.gemspec]: this should be ` s.add_runtime_dependency "logstash-core-plugin-api", ">= 1.60", "<= 2.99"`

File: logstash-filter-kubernetes_metadata.gemspec [hunk 6]
```diff
@@ -0,0 +1,26 @@
+Gem::Specification.new do |s|
+  s.name          = 'logstash-filter-kubernetes_metadata'
+  s.version       = '1.0.2'
+  s.licenses      = ['AGPL-3.0']
```
Comment (suyograo) [inline logstash-filter-kubernetes_metadata.gemspec]: Why not Apache 2.0 license?

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+## 0.1.0
```
Comment (suyograo) [inline CHANGELOG.md]: Version here does not match version in gemspec

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 8]
```diff
@@ -0,0 +1,223 @@
+# encoding: utf-8
+require "logstash/filters/base"
+require "logstash/namespace"
+require "lru_redux"
+require "rest-client"
+require "uri"
+require "logstash/json"
+
+# Process the metadata include in the file names of kubernetes log files
```
Comment (suyograo) [inline lib/logstash/filters/kubernetes_metadata.rb]: Can you add more information here? This will be converted to user facing documentaion.

**Comment by phutchins** [comment]
Working on the requested updates. Will have addressed shortly.

File: Gemfile.lock (hunk 1 above)
Comment (phutchins) [inline Gemfile.lock]: Removed and added it to the gitignore file.

File: CHANGELOG.md (hunk 7 above)
Comment (phutchins) [inline CHANGELOG.md]: Updated to match.

File: CHANGELOG.md (hunk 2 above)
Comment (phutchins) [inline CHANGELOG.md]: Updated changelog message. As this would be the initial release I simply stated what the plugin is/does.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 8 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Updated with simple documentation on how to use and configure the plugin. I can go into more detail but this should be enough to get us going.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 3 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Updated this and everything else to be in asciidoc format.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 4 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Updated.

File: logstash-filter-kubernetes_metadata.gemspec (hunk 6 above)
Comment (phutchins) [inline logstash-filter-kubernetes_metadata.gemspec]: I generally use AGPL-3 on most of my code but don't mind flipping it to Apache 2.0. Updated.

File: logstash-filter-kubernetes_metadata.gemspec (hunk 5 above)
Comment (phutchins) [inline logstash-filter-kubernetes_metadata.gemspec]: Updated.

**Comment by phutchins** [comment]
Changes completed.

**Comment by phutchins** [comment]
Added a more detailed description in the beginning of the file.

**Comment by phutchins** [comment]
@suyograo, have I addressed all of your requested changes?

**Comment by phutchins** [comment]
bump...

**Comment by phutchins** [comment]
@suyograo we still planning on merging this?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+# {
+#   "path" => "/logs/kube-logs/ssl-manager-535881469-x6hmm_storj-prod_ssl-manager-c817d2905d339677288ff73375856a066d4b4d8d45482e1f2e234428d217eb19.log",
+#   "kubernetes" => {
+#     "log_format_stdout" => "default",
+#     "pod" => "ssl-manager-535881469-x6hmm",
+#     "container_name" => "ssl-manager",
+#     "log_format_stderr" => "default",
+#     "namespace" => "storj-prod",
+#     "replication_controller" => "ssl-manager-535881469",
+#     "annotations" => {
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: json embedded within a string?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+#     "labels" => {
+#       "app" => "ssl-manager",
+#       "pod-template-hash" => "535881469",
+#       "version" => "latest"
+#     }
+#   },
+#   "@timestamp" => 2017-02-13T02:33:11.702Z,
+#   "@version" => "1",
+#   "host" => "floptop.local",
+#   "message" => "{\"log\":\"10.244.1.1 - - [25/Jan/2017:14:51:38 +0000] \\\"GET / HTTP/1.1\\\" 200 0 \\\"-\\\" \\\"GoogleHC/1.0\\\" \\\"-\\\"\\n\",\"stream\":\"stdout\",\"time\":\"2017-01-25T14:51:38.65719468Z\"}",
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: json embedded in a string?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+      rest_opts = {
+        verify_ssl: false
+      }
+
+      if @auth
+        if @auth['basic']
+          @logger.debug("Found basic auth for Kubernetes API")
+
+          basic_user = @auth['basic']['user']
+          basic_pass = @auth['basic']['pass']
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Can you add extra validation here for these settings? It would be easy for a user to typo 'user' and 'pass' (and 'basic' and 'bearer', etc) and they won't get any meaningful feedback from the configuration validator about this.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+  def get_kubernetes(namespace, pod)
+
+    begin
+      @logger.debug("Attempting to query the Kubernetes API")
+
+      url = [ @api, 'api/v1/namespaces', namespace, 'pods', pod ].join("/")
+
+      rest_opts = {
+        verify_ssl: false
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: this is a bad default. Disabling SSL verification takes the "security" out of SSL.

Why disable SSL verification?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+      if response && response.code != 200
+        @logger.warn("Non 200 response code returned: #{response.code}")
+      end
+
+      @logger.debug("response was: #{response}")
+
+      data = LogStash::Json.load(response.body)
+
+      {
+        'annotations' => sanatize_keys(data['metadata']['annotations']),
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: typo. sanatize -> sanitize

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        @logger.warn("url: #{url}, rest_opts: #{rest_opts}")
+      rescue Exception => e
+        @logger.warn("Error while querying the API: #{e.to_s}")
+      end
+
+      if response && response.code != 200
+        @logger.warn("Non 200 response code returned: #{response.code}")
+      end
+
+      @logger.debug("response was: #{response}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging is available for use:

```
@logger.debug("Got response from Kubernetes API", response: response)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        response = RestClient::Resource.new(url, rest_opts).get
+      rescue RestClient::ResourceNotFound
+        @logger.warn("Kubernetes returned an error while querying the API")
+        @logger.warn("url: #{url}, rest_opts: #{rest_opts}")
+      rescue Exception => e
+        @logger.warn("Error while querying the API: #{e.to_s}")
+      end
+
+      if response && response.code != 200
+        @logger.warn("Non 200 response code returned: #{response.code}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: use structured logging:

```
@logger.warn("Non 200 response code returned.", code: response.code)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+          rest_opts.merge!( Authorization: "Bearer #{bearer_key}" )
+        end
+      end
+
+      @logger.debug("rest_opts: #{rest_opts}")
+
+      begin
+        response = RestClient::Resource.new(url, rest_opts).get
+      rescue RestClient::ResourceNotFound
+        @logger.warn("Kubernetes returned an error while querying the API")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Structured logging:

```
@logger.warn("Kubernetes returned an error while querying the API", url: url)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+        @logger.warn("url: #{url}, rest_opts: #{rest_opts}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: I recommend not logging `rest_opts` because it can contain secrets (user, password, tokens, etc), and we should not log secrets.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+      rescue Exception => e
+        @logger.warn("Error while querying the API: #{e.to_s}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging:

```
@logger.warn("Error while querying the API", :error => e.to_s)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+      @logger.debug("response was: #{response}")
+
+      data = LogStash::Json.load(response.body)
+
+      {
+        'annotations' => sanatize_keys(data['metadata']['annotations']),
+        'labels' => sanatize_keys(data['metadata']['labels'])
+      }
+    rescue => e
+      @logger.warn("Unknown error while getting Kubernetes metadata: #{e}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging

```
@logger.warn("Unknown error while getting Kubernetes metadata", :error => e.to_s)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+          if v = a[k]
+            format[t] = v
+            break
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata => #{metadata}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging

```
@logger.debug("Kubernetes metadata", metadata: metadata)
```

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+
+    rescue => e
+      @logger.warn("Error setting log format: #{e}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging

```
@logger.warn("Error setting log format", error => e.to_s)
```

Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: For warnings, what action wil the user take here? Is "error setting log format" informative enough for a user to know what action to take?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+    self.lookup_cache ||= LruRedux::ThreadSafeCache.new(1000, 900)
+    @logger.debug("Created cache...")
+  end
+
+  # this is optimized for the single container case. it caches based on filename to avoid the
+  # filename munging on every event.
+
+  public
+  def filter(event)
+    @logger.debug("event is: #{event}")
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: structured logging

```
@logger.debug("event", event => event)
```

But maybe you don't need this. Logstash pipeline will already log (in debug mode, I think) that a filter is receiving something like this.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+  #   }
+  # }
+  # -----------------------------------------------------------------
+  config :auth, :validate => :hash, :default => {}
+
+  # Kubernetes API URL
+  config :api, :validate => :string, :default => "http://127.0.0.1:8001"
+
+  # Default log format
+  config :default_log_format, :validate => :string, :default => "default"
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Is the docs here enough information for a user to decide? What other log format values would there be?

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 9 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: This particular annotation is returned from Kubernetes as an escaped JSON string. Not all annotations are returned this way. I feel its best to leave it up to the user to decide if they want to continue parsing JSON from within a particular annotation.

If this pod had had other annotations that were non JSON strings, they would have been displayed as such.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 10 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Similar to above, I don't want to assume that every app running inside kubernetes will be logging in JSON. I leave it up to the user to use other filters and flags to determine if they should further parse the message.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 23 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Definitely not. Pushed a change with better documentation. :)

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 22 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Removed this one. We can add it back if it turns out the pipeline doesn't already log it.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 20 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 21 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Added a little more detail. Hard to be too specific but this should help.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 12 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Added a config param for this and defaulted to true.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 11 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Moved to a flat auth_basic_user, etc... approach and using built in config validation which should work better.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 16 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 17 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Removed rest_opts.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 18 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 15 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 14 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 13 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Whoops! Fixed. :)

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 19 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Fixed.

**Comment by phutchins** [comment]
Thanks for the review @jordansissel. I believe I've addressed all changes requested. Let me know if anything still needs tweaking.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 9 above)
Comment (chancez) [inline lib/logstash/filters/kubernetes_metadata.rb]: I agree and can also verify what @phutchins says. One approach you could take is to add a Boolean configuration option which checks for the first character being { and last being }, and attempt to parse as json optimistically, the fluentd version of the filter does this. Hard to say if it's a good approach though.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+      if response && response.code != 200
+        @logger.warn("Non 200 response code returned", code: response.code)
+      end
+
+      @logger.debug("Got response from Kubernetes API", response: response)
+
+      data = LogStash::Json.load(response.body)
+
+      {
+        'annotations' => sanitize_keys(data['metadata']['annotations']),
```
Comment (chancez) [inline lib/logstash/filters/kubernetes_metadata.rb]: You probably should make annotations optional since they often include vary wide of things hat people won't always want collected in their logs. And some components may have a lot of annotations, not all of which are useful and may just bloat the message (last applied config is a good example) Probably best to allow users to whitelist keys.

**Comment by jordansissel** [comment]
@phutchins sweet. I'll review this again shortly and have put it on my todo list for this week :)

**Comment by phutchins** [comment]
@jordansissel awesome, sounds good.

@chancez thats a good idea. I'll take a look at implementing a default on and empty whitelist with an optional flag to turn it off and allow all annotations which should catch all use cases. Will also look into JSON parsing. I'm curious what the benchmarks would look like between attempting to parse the JSON with a catch in case of failure.

I've done this in a ruby block in one of my logstash configs (code included below)... It handles nested json as well but I'd want to be aware of the performance cost before adding it here...

```
  if [log_format] == "json" {
    ruby  {
      init => "require 'json'"
      code => "
        def valid_json?(json)
          begin
            parsed_json = JSON.parse(json)
            return parsed_json
          rescue
            return false
          end
        end

        e = event.get('message')

        parsed_json = valid_json?(e)

        if parsed_json
          parsed_json.each do |field_name, field_val|
            if field_name == 'message'
              field_name = 'message_field'
            end

            event.set(field_name, field_val)
          end

          # Try to pull out valid json from the hash['message'] string
          if parsed_json['message']
            matches = /(?<!\{)[^\{]*(\{.*\})(?!\})/.match(parsed_json['message'])

            if matches && matches[1]
              nested_json = valid_json?(matches[1])
              log_message = nested_json ? nested_json : matches[1]
            end

            if !nested_json
              event.tag('_json_parse_failure')
            end

            if log_message
              event.set('log_message', log_message)
            end
          end
        end
      "
    }
  }
```

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 10 above)
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Ahh I understand now. This is just the content of the log file.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+  config :auth_basic_pass, :validate => :string
+
+  # Bearer Auth Key
+  config :auth_bearer_key, :validate => :string
+
+  # Kubernetes API URL
+  config :api, :validate => :string, :default => "http://127.0.0.1:8001"
+
+  # Verify Kubernetes API SSL
+  config :verify_api_ssl, :validate => :boolean, :default => true
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Do we need this setting? Disabling SSL verification means disabling the reason we choose to have SSL in the first place :\

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+
+  def set_log_formats(metadata)
+    begin
+      #return if metadata['annotations'].empty?
+
+      format = {
+        'stderr' => @default_log_format,
+        'stdout' => @default_log_format
+      }
+      a = metadata['annotations']
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: These variable names are very short and it makes it hard to follow the code with so many single-letter names.

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata", metadata: metadata)
+
+    rescue => e
+      @logger.warn("Error setting log format. Please check log-format annotation.", error => e.to_s)
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: Should this tag the event as having failed?

File: lib/logstash/filters/kubernetes_metadata.rb [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+  config_name "kubernetes_metadata"
+
+  # The source field name which contains full path to kubelet log file.
+  config :source, :validate => :string, :default => "path"
+
+  # The target field name to write event kubernetes metadata.
+  config :target, :validate => :string, :default => "kubernetes"
+
+  # Auth for hitting the Kubernetes API. This can be either basic auth or
+  # Bearer auth. If you specify both, it will default to basic.
```
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: For user experience, this means that if they make a mistake and provide both, the plugin appears to silently decide to use basic auth.

Options:
* Can you make this a failure if users provide both basic and bearer auth?
* ALternately, at least warn the user if they provide both.

File: spec/filters/kubernetes_metadata_spec.rb [hunk 29]
```diff
@@ -0,0 +1,363 @@
...
+require "spec_helper"
+require "logstash/filters/kubernetes_metadata"
+require "logstash/json"
+require "sinatra"
+require "thread"
+
+PORT = rand(65535-1024) + 1025
+
+
+class TestApp < Sinatra::Base
```
Comment (jordansissel) [inline spec/filters/kubernetes_metadata_spec.rb]: I love this test fixture. Nice work :)

**Comment by chancez** [comment]
Also, have you considered using the kubeclient gem? It would allow you to use efficient watching of the API for changes (to annotations/labels), so that your cache data can be updated when things are updated in the API. Could be done in a follow up PR later however.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 28 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Sure thing. I went with warning the user and defaulting to basic for now. I can change this to a failure if you can point me to a good example of how we should fail on config issues.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 25 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Generally I would agree as without verification there is far less value in using SSL to begin with. In this case however, the Kubernetes API comes configured (when created with the provided kube-up.sh script, and other methods) out of the box with SSL configured to verify against its own internal CA. 

Without adding a valid public cert or configuring logstash to use the kubernetes CA, we're stuck with no SSL/TLS validation or communicating in the open. I'd much prefer using unvalidated SSL over plain text communication. So in this case I do think that this setting is reasonable to have as an option.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 26 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Totally agree. Updated with more descriptive variable names.

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 27 above)
Comment (phutchins) [inline lib/logstash/filters/kubernetes_metadata.rb]: Probably would be good to know. I went ahead and added a `_kubelogformatparsefailure` tag before this warn.

File: spec/filters/kubernetes_metadata_spec.rb (hunk 29 above)
Comment (phutchins) [inline spec/filters/kubernetes_metadata_spec.rb]: I do too! I would definitely like to add more test coverage in the future... Thanks. :)

**Comment by phutchins** [comment]
All of @jordansissel's requested changes have been addressed.

@chancez using the downward API is a great idea. Along with this, I'd also like to include a dockerfile for using logstash this way and some example configs for kubernetes...

Currently I'm running Filebeat in kubernetes so that I can use TLS, which ships to an external logstash. I've been hesitant to move my ES cluster to Kubernetes as when things break it's not always the best idea to have all of your eggs in one basket... Especially logging as it's generally the first place you go to begin troubleshooting.

I'd not actually discovered the kubeclient gem yet. That sounds like it could make things a lot more efficient. I'll definitely look into adding this down the road. Thanks for the awesome ideas!

**Comment by phutchins** [comment]
The last thing I need to do is to make the annotations whilelist feature and this should be ready to go. Should be done with this shortly.

**Comment by phutchins** [comment]
Added whitelist and tests to cover. Later we could possibly add an array of annotations that should be parsed as JSON or if attempting to parse all annotations as JSON is performant enough, we could do that. 

Should be ready to go!

File: lib/logstash/filters/kubernetes_metadata.rb (hunk 9 above)
Comment (jordansissel) [inline lib/logstash/filters/kubernetes_metadata.rb]: This helps me understand, thank you!

**Comment by jordansissel** [comment]
Thank you for this wonderful work :)

@suyograo, did you still have outstanding change items?
